Vostok.ClusterClient.Core/Modules/RequestContext.cs

```diff
@@ -48,7 +52,29 @@ internal class RequestContext : IRequestContext
 
         public ILog Log { get; }
 
-        public IClusterProvider ClusterProvider { get; set; }
+        public IClusterProvider ClusterProvider
+        {
+            // note (kungurtsev, 26.05.2022): should be not null
+            get =>
+                asyncClusterProvider != null
+                    ? new SyncClusterProviderAdapter(asyncClusterProvider)
+                    : clusterProvider;
+            set
+            {
+                clusterProvider = value;
+                asyncClusterProvider = null;
+            }
+        }
+        public IAsyncClusterProvider AsyncClusterProvider
+        {
+            // note (kungurtsev, 26.05.2022): may be null, then ClusterProvider not null
```

Comment:
What's the point of having nullable `IAsyncClusterProvider`?
Why not just use `AsyncClusterProviderAdapter`?

Reply:
To easily check which one is better for use ([RequestExecutionModule.cs](https://github.com/vostok/clusterclient.core/pull/35/files#diff-8f316f7e9382c3b92d15bb1cce8d0a56dbb06a6630fb9f8c6b0c00b81bf5806c) and avoid extra async calls.

Reply:
Looks like a tideous work to do.
Also, afair, the only overhead of extra async call which is synchronous internally is the allocation of `Task` class.
No jobs are scheduled to the thread pool because of the absence of actual 'await' calls.

I also did a small benchmark: 
```
[Benchmark]
public void SyncTest()
{
    Job();
}

[Benchmark]
public async Task AsyncTest()
{
    await AsyncJob();
}

public static Task<int> AsyncJob()
{
    return Task.FromResult(Job());
}

public static int Job()
{
    var i = 0;
    for (var j = 0; j < 100000; ++j)
        i += j;
    return i;
}
```

Results:
```
|    Method |     Mean |    Error |   StdDev | Allocated |
|---------- |---------:|---------:|---------:|----------:|
|  SyncTest | 24.99 us | 0.219 us | 0.195 us |         - |
| AsyncTest | 25.29 us | 0.425 us | 0.473 us |      72 B |
```

Are you sure this is worth it?

Reply:
But actual Job probably is much more faster than your:
```
private static readonly Uri[] urls = Enumerable.Range(0, 10).Select(i => new Uri($"http://kontur.ru/{i}")).ToArray();
public static Uri[] Job()
{
    return urls;
}
```

Results:
```
|    Method |       Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------- |-----------:|----------:|----------:|-------:|------:|------:|----------:|
|  SyncTest |  0.2386 ns | 0.0502 ns | 0.0470 ns |      - |     - |     - |         - |
| AsyncTest | 26.4652 ns | 0.5811 ns | 0.6692 ns | 0.0344 |     - |     - |     144 B |
```

Reply:
Anyway, this can be easily changed later.

Reply:
Also you can compare TransformingAsyncClusterProvider with TransformingClusterProvider.

I prefer to have both of them and never mix sync and async pipeline.

Reply:
Ok, let's keep it as it is